File: learning_ros2/src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
# Allows us to add more arguments to our callback
from functools import partial
 
 
class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.call_add_two_ints_server(6, 7)
        # You can call them mutliple times for here
        #self.call_add_two_ints_server(4, 3)

    ### Start of template
    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server Add Two Ints...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        future = client.call_async(request)
        # This add done callback is part of the future method and has nothing to do with
        # ROS. Here we use it since in python we can't wait asynchronously without
        # multi-threaded operation
        future.add_done_callback(partial(self.callback_call_add_two_ints, a=a, b=b))

    def callback_call_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(str(a) + " + " + str(b) + " = " +
                    str(response.sum))
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))
    ### End of template

 
def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: ROS2/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClientNode : public rclcpp::Node
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client")
    {
        //callAddTwoIntsService(1, 4);

        // If you call the function directly: the program starts -> we start the node
        // -> we create the client -> we wait for service -> if we start the server on
        // another terminal it passes -> we send the request -> we'll wait for the
        // response at future.get (problem) -> This will block the thread here
        // -> which means the function will not exit -> so the constructor will
        // not exit -> so we'll still be in this line:
        // "auto node = std::make_shared<AddTwoIntsClientNode>();" -> and spin(node)
        // will not be executed. -> But we need spin(node) in order to get the result
        // from the future.
        // So we need to start this in a different thread so that we can continue the
        // execution, and the constructor can exit and we can call spin on the node.
        // So we create a thread object in private
        thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4));

        // If you'd like to send multiple requests you can use a thread pool
        //threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4)));
        //threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 2, 5)));

    }

    // This function needs to be called on a different thread
    void callAddTwoIntsService(int a, int b)
    {
        // Create client with type and name
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        // Wait for service to be up
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server to be up...");
        }

        // We fill the request with what we need and send it
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        // This will block the thread
        auto future = client->async_send_request(request);

        try
        {
            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %d", a, b, response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        }
    }

private:
    std::thread thread1_;
    // Example solution for multiple requests
    //std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
